x-pack/plugins/alerting/server/rules_client/lib/validate_rule_aggregation_fields.ts

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import Boom from '@hapi/boom';
+import type { AggregationsAggregationContainer } from '@elastic/elasticsearch/lib/api/typesWithBodyKey';
+
+const ALLOW_FIELDS = [
+  'alert.attributes.executionStatus.status',
+  'alert.attributes.lastRun.outcome',
+  'alert.attributes.muteAll',
+  'alert.attributes.tags',
+  'alert.attributes.snoozeSchedule',
+  'alert.attributes.snoozeSchedule.duration',
+  'alert.attributes.alertTypeId',
+  'alert.attributes.enabled',
+  'alert.attributes.params.*',
+];
+
+const ALLOW_AGG_TYPES = ['terms', 'composite', 'nested', 'filter'];
+
+const aggTypesoVerify = ['field', 'path'];
+
+const aggregationKeys = ['aggs', 'aggregations'];
+
+export const validateRuleAggregationFields = (
+  aggs: Record<string, AggregationsAggregationContainer>
+) => {
+  Object.values(aggs).forEach((aggContainer) => {
+    // validate root level aggregation types (non aggs/aggregations)
+    validateTypes(aggContainer);
+
+    // Recursively go through aggs to validate terms
+    if (aggContainer.aggs) {
```

Comment:
> Is empty (containing undefined) aggs a valid field? This part could be simplified by accepting undefined as well as a valid value. This way it won't interfere with aggregationKeys as it seems like a duplication right now.

It shouldn't be valid, but there's built-in validation by the savedobject client that will validate the undefined case. This validation is simply to validate terms against our specific allow list.

> If there is no difference between aggs and aggregations rules client could normalize requests as well to use one field. What if both fields are given?

aggs and aggregations are both valid inputs, If both fields are given they will both be validated since we don't return immediately once we encounter either

Reply:
I see. So you delegate aggregation query object validation to the SO client and don't want to introduce too much complexity here to improve maintainability. It looks good to me.

And one small thing here. Lines `37`, `41` and `49` look connected. Does it look right to you modify `validateRuleAggregationFields` a bit to make it like this

```
export const validateRuleAggregationFields = (
  aggs: Record<string, AggregationsAggregationContainer>
) => {
  Object.values(aggs).forEach((aggContainer) => {
    // Recursively go through aggs to validate terms
    if (aggContainer.aggs) {
      validateRuleAggregationFields(aggContainer.aggs);
      return;
    }
    if (aggContainer.aggregations) {
      validateRuleAggregationFields(aggContainer.aggregations);
      return;
    }

    validateTypes(aggContainer);
  });
};
``` 

so `AGG_KEYS` constant can be removed as well as lines `48`-`51`?

Reply:
Unfortunately if I return early it fails tests like these: 

```ts
    expect(() => {
      validateRuleAggregationFields({
        name1: {
          terms: {
            field: 'alert.attributes.snoozeSchedule',
          },
          aggs: {
            nestedAggs1: {
              terms: {
                field: 'alert.attributes.enabled',
              },
            },
          },
          aggregations: {
            nestedAggs2: {
              terms: {
                field: 'alert.attributes.consumer',
              },
            },
          },
        },
      });
    }).toThrowErrorMatchingInlineSnapshot(`"Invalid aggregation term: alert.attributes.consumer"`);
```
since it returns the first aggs/aggregations it sees at any particular nesting level